strax/storage/mongo.py

```diff
@@ -133,46 +136,81 @@ def __init__(self, key, metadata, col):
+        self.id_md = self.col.insert_one(meta_data).inserted_id
+        # Also save all the chunks
+        self.ids_chunk = {}
 
-    def _save_chunk(self, data, chunk_info, executor):
+    def _save_chunk(self, data, chunk_info, executor=None):
         """see strax.Saver"""
         chunk_number = chunk_info['chunk_i']
+        chunk_name = f'chunk_{chunk_number}'
 
+        aggregate_data = []
         # Remove the numpy structures and parse the data. The dtype information
         # is saved with the metadata so don't worry
         for row in data:
             ins = {}
             for key in list(data.dtype.names):
                 ins[key] = row[key]
             ins = remove_np(ins)
-            self.col.update_one({'_id': self.id},
-                                {'$addToSet': {f'chunk_{chunk_number}':
-                                                   {"data": ins}}
-                                 })
+            aggregate_data.append(ins)
+
+        # Get the document to update, if none available start a new one for
+        # this chunk
+        chunk_id = self.ids_chunk.get(chunk_name, None)
+        if chunk_id is not None:
+            self.col.update_one({'_id': chunk_id},
+                                {'$addToSet': {f'chunk_name.data': aggregate_data}})
+        else:
+            # Start a new document, update it with the proper information
+            doc = self.basic_md.copy()
+            doc['write_time'] = datetime.now(py_utc)
+            doc[chunk_name] = {"data": aggregate_data}
+
+            chunk_id = self.col.insert_one(doc).inserted_id
+            self.ids_chunk[chunk_name] = chunk_id
+
         return dict(), None
 
     def _save_chunk_metadata(self, chunk_info):
         """see strax.Saver"""
-        # For the first chunk we update the run_start_time
+        # For the first chunk we get the run_start_time and update the run-metadata file
         if int(chunk_info['chunk_i']) == 0:
-            t0 = datetime.fromtimestamp(chunk_info['start']/1e9).replace(tzinfo=py_utc)
-            self.col.update_one({'_id': self.id},
+            self.run_start = t0 = datetime.fromtimestamp(
+                chunk_info['start']/1e9).replace(tzinfo=py_utc)
+            self.col.update_one({'_id': self.id_md},
                                 {'$set':
                                      {'run_start_time': t0}})
 
-        self.col.update_one({'_id': self.id},
+        self.col.update_one({'_id': self.id_md},
                             {'$addToSet':
                                  {'metadata.chunks': chunk_info}})
 
     def _close(self):
         """see strax.Saver"""
+        # First update the run-starts of all of the chunk-documents as this is
+        # a TTL index -candidate
+        if self.run_start is not None:
+            update = {'run_start_time': self.run_start}
+            for chunk_id in self.ids_chunk.keys():
+                self.col.update_one({'_id': chunk_id}, {'$set': update})
```

Comment:
Probably faster to do one call to `update_many({'_id': {'$in': list(self.ids_chunk.keys())}}, ...)` than many calls to `update_one`. Also probably possible to just specify `{'number': run_number}`.

Reply:
Thank, good catch